### ISSUE TITLE
Production usecase/instructions

### ISSUE THREAD

**Issue by erbilsilik** [grumpyTofu/nx-mongo-migrate#7]
**Is your feature request related to a problem? Please describe.**
The package performs effectively when executed in a local environment. However, there is a lack of explicit guidance regarding its implementation in a production setting. Furthermore, the published version is not functioning as intended due to its dependency on a `migration.config.ts` (TypeScript) file, which it is unable to locate in the specified path.

**Describe the solution you'd like**
By using [Kubernetes init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) - This is the preferred approach. By triggering the executor from the built image, we can execute this image prior to running the application.

**Describe alternatives you've considered**
By using CI/CD pipelines - `npx nx affected --target=migrate-up ` While it is technically feasible to utilize the solution through the pipeline, I would not recommend this approach as it necessitates granting production credentials to the Continuous Integration (CI) system.

**Additional context**

**Comment by grumpyTofu** [comment]
Hey @erbilsilik, you are correct on the lack of guidance ;) I was looking to allow other devs to determine their own implementation and alter the package to accommodate any changes required as needed.

That being said, the approach that I have taken with this library in a production setting was to use `npx nx affected --target=migrate-up`, as you mentioned, but on a private runner. CI/CD is going to the quickest and easiest solution. You should only need to configure your pipeline with the appropriate secrets. I think it should also be able to tap into your cloud provider CLI . You could modify the migration config to be an async function and it should be able to pull your db connection url from your chosen secret manager solution. If you ultimately choose to go this route and it does not work, let me know and we should be able to modify the library to work with this implementation.

It sounds like the issue you are experiencing is that the nx build is not including your migration config in your build and the dockerfile is not copying the migration config into the image. Fixing this might require some tinkering, but should be possible. We can also modify the library to provide any utility commands to make this easier. Another option would be to move all the project specific stuff under the src directory and let it build with the app. I think the library might need to be modified to make that work but I am not sure without trying it. I originally opted out of this idea, because I didn't want all the migrations bundled with my app if they were not needed by the app, but that is of course because I ran it in CI/CD.

If you are looking to use this with docker/k8, my only caution would be that you may potentially run into a race condition trying to apply migrations. The idea you have for the init containers seems interesting though. To be honest, I am not the most experienced with k8. But, it sounds like it would be really cool to support this out of the box. So, if you have an idea how this could be implemented, I would be open to whatever your suggestion/recommendation is here.

**Comment by erbilsilik** [comment]
Hey @grumpyTofu  ,

Thank you for your prompt response.

Indeed, implementing CI/CD (Continuous Integration/Continuous Deployment) is a suitable solution for the current setup. However, there is another consideration to keep in mind apart from accessing PROD (production) credentials through CI/CD. If you are using an Infrastructure as Code (IaC) approach to manage your infrastructure, you will need to run migrations before validating the Kubernetes resources.

You can easily reproduce the issue yourself by creating a new Nx project and installing this package. To address this problem, I have decided to create a new plugin within my project directory. This allows me to reference my internal library and utilize it within the application. It's important to note that there isn't a one-size-fits-all solution, but I would recommend exposing a similar functionality to NestJs's useFactory function.

Recently, I introduced a migration configuration per app, which helps to prevent any race conditions since each app will have its own database. This approach is not directly related to Kubernetes (k8s), but rather involves building migrations.ts files alongside the app and using the same image to execute the migrations. Let me provide an example to illustrate this: Suppose I have an application that I transpile to JavaScript using @nrwl/node:webpack, with the main entry point and migration files located within the same app/migrations directory. Additionally, let's say I created this package under the **libs** directory. In such a scenario, what steps do I need to take in order to run the migrations first and then start the application? Once we achieve this, we can easily incorporate it into a Kubernetes environment.

I hope this clarifies the situation.

**Comment by grumpyTofu** [comment]
I am glad to hear that the recent changes have enabled you to move forward with utilizing the library for your app. The idea of utilizing the library inside of a nest app also sounds interesting. If you feel that it is a valuable addition for others and would mesh well with this library, feel free to open a PR!

Nice! if the database per app fixes the potential race condition, then that is something that should probably be added to documentation. That way others are aware that utilizing this in a k8 environment with a shared database could lead to a potential race condition when applying migrations.

Now, as far as this part w/ webpack, I can say that one potential way to solve this issue is a custom webpack config. You can write a multi-stage webpack build that packs the app and migrations separately and customizes the entry points plus outputs based on your specific needs. Here is an example of one similar to what I use. Its not tested, so use at your own peril ;) Again though, if you do go this route and feel that parts of your solution are modular enough to benefit others, please open a PR

```js
const { composePlugins, withNx } = require('@nrwl/webpack');
const { sync } = require('glob');
const { join } = require('path');

const getMigrationEntries = (context) => {
  const metadata = context.workspace.projects[context.projectName];
  const projectRoot = metadata.root;
  const migrationsGlob = join(projectRoot, 'migrations', '*.ts');
  const replaceRegex = new RegExp(join(projectRoot, 'migrations', '(.*[js|ts])'));

  return sync(migrationsGlob).reduce((acc, file) => {
    const step1 = file.replace(replaceRegex, '$1');
    const outputName = step1.replace(/(.*)\.[^.]+$/, '$1');
    return {
      ...acc,
      [join('migrations', outputName)]: file,
    };
  }, {});
};

const getAppConfig = composePlugins(withNx(), (config, { context }) => {
  // whatever custom stuff your app needs for build
  return config;
});

const getMigrationConfig = composePlugins(withNx(), (config, { context }) => {
  return {
    ...config,
    entry: {
      ...config.entry,
      ...getMigrationEntries(context),
    },
  };
});

// Important Note: This can be async
 module.exports = (config, context) => {
  const appConfig = getAppConfig(config, context);
  const migrationConfig = getMigrationConfig(config, context);
  return [appConfig, migrationConfig];
};
```

**Comment by erbilsilik** [comment]
Impressive! 

It seems that the only remaining piece is the `migration.config.ts` file, which I believe we can resolve using the same approach by just adding the following
```
"additionalEntryPoints": [
          {
            "entryName": "migration.config",
            "entryPath": "apps/example/migration.config.ts"
          }
]
```

Additionally, can you provide you with instructions on how to trigger the executor functions (up, status, down) in order to verify if migrations can be performed from the transpired code (**please note**: I want to use the app and lib from **dist** directory)

Regarding the race condition issue, allow me to clarify. By executing the command `npx nx affected --target=migrate-up` the migrations will be executed sequentially, preventing any race conditions. Is that the reason? If that is the case, yes, having database per app will solve the issue since it will access only its own database. Or do you mean something else?

Once I have everything set up correctly, I will be making changes and updating the Readme accordingly and will ensure that others can easily follow the instructions and understand the project's configuration.

**Comment by grumpyTofu** [comment]
AH, yes this is a really good question. I think this is something I have not considered as I was expecting the migrations to be run with Nx. Unfortunately, using this webpack approach to pack up both your apps as one will potentially remove your ability to run Nx migrations. I think there are 2 options:

First, I would check out [Nx docs](https://nx.dev/reference/project-configuration#including-package.json-files-as-projects-in-the-graph) for the package.json configuration that you can use. It looks like Nx is suggesting that there is a way to transfer your project.json targets into your package.json during your build process. I don't know what that looks like in the dist build but it could potentially allow you to run the executor on the build. This would be a really nice way of solving the problem if it works.

If that doesn't work, I think the only other way to resolve that issue will be to refactor the executor methods so that the core functionality is abstracted out into a reusable function that can be imported and triggered programmatically from another file like `migrations.ts`. This would become another new entry point for webpack that you need to run the migrations. It would probably need to use yargs or something so you could specify the behavior you want (up, down, status). Let me know if the first one doesn't work, I can take a shot at refactoring the executors hopefully soon.

For the race condition; I think yes, that is mostly correct. Remember though that the nx affected commands are not necessarily run sequentially. They can be parallelized with `nx affected --target=migrate-up --parallel=5`. I believe this is actually the recommended approach if you are using Nx Cloud for performance optimization of your pipelines. I have always had issues with it though, so I try to avoid using it personally. Its probably not a good idea to try and run migrations in parallel.

**Comment by grumpyTofu** [comment]
Oh, no I was not aware. I will open an issue for that one and start investigating that issue as soon as possible. It should be possible to use this library just by installing.

**Comment by grumpyTofu** [comment]
built-in TS bundling added via: PR https://github.com/grumpyTofu/nx-mongo-migrate/issues/9

**Comment by erbilsilik** [comment]
I've talked about this with my team and the consensus is that managing the migration from the pipeline is not a good idea. Do you think we can somehow develop the package that enables to use transpiled version? Maybe by just replacing nx executors with nx commands?

**Comment by grumpyTofu** [comment]
Yes, absolutely! I think refactoring to allow for the core functions to be run from both the executors and from a custom script would solve that issue and would probably give you the most mileage. Although, I think the nx commands should still be able to handle most cases. There shouldn't be anything stopping you from running `npx nx ...command` from any environment and having this work properly.

Unfortunately, I have been a bit swamped with other projects at the moment so I probably won't have a lot of time to work on that upgrade in the near future. If you and your team would find programmatic access to be a beneficial feature, I would more than welcome a PR :)
